py_mini_racer/extension/mini_racer_extension.cc

```diff
@@ -534,91 +559,119 @@ BinaryValue* MiniRacer_eval_context_unsafe(ContextInfo *context_info,
     }
 
-    if (!eval_result.executed) {
+    else if (!eval_result.executed) {
         result = ALLOC(BinaryValue);
         result->type = type_execute_exception;
+        result->str_val = nullptr;
 
-        if(message && TYPE(message) == T_STRING && backtrace && TYPE(backtrace) == T_STRING) {
-            char *msg_str = PSTRING_PTR(message);
-            char *backtrace_str = PSTRING_PTR(backtrace);
-            size_t dest_size = strlen(msg_str) + strlen(backtrace_str) + 1;
-            char *dest = (char *) malloc(dest_size);
+        if (bmessage && bmessage->type == type_str_utf8 &&
+                bbacktrace && bbacktrace->type == type_str_utf8) {
+            // +1 for \n
+            size_t dest_size = bmessage->len + bbacktrace->len + 1;
+            char *dest = static_cast<char *>(malloc(dest_size));
             if (!dest) {
-                BinaryValueFree(message);
-                free(result);
-                return NULL;
+                goto err;
             }
-            snprintf(dest, dest_size, "%s\n%s", msg_str, backtrace_str);
+            memcpy(dest, bmessage->str_val, bmessage->len);
+            dest[bmessage->len] = '\n';
+            memcpy(dest + bmessage->len + 1, bbacktrace->str_val, bbacktrace->len);
 
-            result->value = dest;
+            result->str_val = dest;
             result->len = dest_size;
-        } else if(message && TYPE(message) == T_STRING) {
-            result->value = strdup(PSTRING_PTR(message));
-            result->len = strlen((char *) result->value);
+        } else if(bmessage && bmessage->type == type_str_utf8) {
+            // canibalize bmessage
+            result->str_val = bmessage->str_val;
+            result->len = bmessage->len;
+            free(bmessage);
+            bmessage = NULL;
         } else {
-            result->value = strdup("Unknown JavaScript error during execution");
-            result->len = strlen((char *) result->value);
+            result->str_val = strdup("Unknown JavaScript error during execution");
+            result->len = result->str_val ? strlen(result->str_val) : 0;
         }
-        BinaryValueFree(message);
-        BinaryValueFree(backtrace);
-
-        return result;
     }
-    BinaryValueFree(message);
-    BinaryValueFree(backtrace);
 
     // New scope for return value
-    {
+    else {
         Locker lock(context_info->isolate);
         Isolate::Scope isolate_scope(context_info->isolate);
         HandleScope handle_scope(context_info->isolate);
 
         Local<Value> tmp = Local<Value>::New(context_info->isolate, *eval_result.value);
         result = convert_v8_to_binary(context_info, tmp);
+    }
 
-        eval_result.value->Reset();
-        delete eval_result.value;
+    if (false) {
+err:
+        BinaryValueFree(result);
+        result = nullptr;
     }
 
+    BinaryValueFree(bmessage);
+    BinaryValueFree(bbacktrace);
+
     return result;
 }
 
+class BufferOutputStream: public OutputStream {
```

Comment:
Should we mark this code as explicitly for debugging purposes? It's on purpose massively leaky.